elastic/kibana
# [Response Ops] [Alerting] Adds circuit breaker that limits the max number of alerts that can be reported each rule run.

Resolves https://github.com/elastic/kibana/issues/138308

## Summary

Adds alert circuit breaker that limits the number of alerts that can be reported during each rule execution.
* Config added under `xpack.alerting.rules.run.alerts.max`. Defaults to 1000
* Updated the alert factory to set a `hasReachedAlertLimit` flag when the limit is reached and throw an error
* Updated `processAlerts` function take `hasReachedAlertLimit` flag and change the way alerts are processed accordingly
  * Any existing alerts from the previous execution of the rule are persisted to the next execution
  * No recovered alert determination is made
  * New alerts are only added up to the capacity determined by `maxAlerts - # existing alerts`
* Update rule execution status to `warning` if the alert limit has been reached. This automatically shows up in the UI:
<img width="2277" alt="Screen Shot 2022-08-10 at 4 45 42 PM" src="https://user-images.githubusercontent.com/13104637/184258573-2ef51fbb-73c6-4bf7-9b7e-e6f078dc0894.png">

## To Verify

1. Set the limit lower in the `kibana.yml` file (I used 20)
2. I used [`es-apm-sys-sim`](https://github.com/pmuellr/es-apm-sys-sim) to start generating data. I started with 5 hosts
3. Create an index threshold rule that groups by host name and always alerts (CPU pct > 0). Attach a server log action that notifies on every active alert.
4. See rule run and generate 5 alerts.
5. Re-run `es-apm-sys-sim` with 30 hosts.
6. The next run of the rule should hit the circuit breaker. You should see the warning in the UI and the number of alerts should max out at 20

### Checklist
